Websocket state errors with production usage

### Description

Using this in production, we often get websocket state errors that lead to us having to take the connection down. Here are is an example (not the frames may not be in order due to how our logs are processed):
```
[ERROR] @slack/client:RTMClient A websocket error occurred: Unexpected server response: 503 
Error: Unhandled event 'websocket close' in state 'connecting'. 
    at /home/node_modules/finity/lib/core/HierarchicalStateMachine.js:79:33 
    at StateMachine.handleUnhandledEvent (/home/node_modules/finity/lib/core/StateMachine.js:76:13) 
    at TaskScheduler.execute (/home/node_modules/finity/lib/core/TaskScheduler.js:29:7) 
    at WebSocket.websocket.addEventListener.event (/home/node_modules/@slack/client/dist/RTMClient.js:391:77) 
    at HierarchicalStateMachine.handle (/home/node_modules/finity/lib/core/HierarchicalStateMachine.js:72:24) 
    at WebSocket.emit (events.js:191:7) 
    at TaskScheduler.enqueue (/home/node_modules/finity/lib/core/TaskScheduler.js:19:12) 
    at emitTwo (events.js:106:13) 
    at WebSocket.onClose (/home/node_modules/ws/lib/event-target.js:124:16) 
    at WebSocket.finalize (/home/node_modules/ws/lib/websocket.js:187:12) 
```
We are also regularly seeing this error appear in logs:
[ERROR] @slack/client:KeepAlive Unhandled error: Failed to send message on websocket: WebSocket is not open: readyState 2 (CLOSING). Please report to @slack/client package maintainers.

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [X] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [ ] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [X] I've read and understood the [Contributing guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [X] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [X] I've searched for any related issues and avoided creating a duplicate issue.

---

### Bug Report

Filling out the following details about bugs will help us solve your issue sooner.

#### Reproducible in:

`@slack/client` version: 4.2.0

node version: v6.10.3

OS version(s): alpine linux 3.7

#### Steps to reproduce:
Haven't been able to figure out specific repro steps unfortunately. Just these observations from production logs

#### Expected result:
No error

#### Actual result:
Errors

#### Attachments:
N/A

--- a/src/RTMClient.ts
+++ b/src/RTMClient.ts
@@ -144,18 +144,25 @@ export class RTMClient extends EventEmitter {
               });
             })
             .on('websocket open').transitionTo('handshaking')
+            .on('websocket close')
+              .transitionTo('reconnecting').withCondition(() => this.autoReconnect)
+              .transitionTo('disconnected').withAction(() => {
+                // this transition circumvents the 'disconnecting' state (since the websocket is already closed), so we
+                // need to execute its onExit behavior here.
+                this.teardownWebsocket();
+              })
           .state('handshaking') // a state in which to wait until the 'server hello' event
             .on('websocket close')
               .transitionTo('reconnecting').withCondition(() => this.autoReconnect)
               .withAction((_from, _to, context) => {
-                this.logger.debug(`reconnecting after unexpected close ${context.eventPayload.reason} 
-                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring} 
+                this.logger.debug(`reconnecting after unexpected close ${context.eventPayload.reason}
+                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring}
                   and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
               })
               .transitionTo('disconnected')
               .withAction((_from, _to, context) => {
-                this.logger.debug(`disconnected after unexpected close ${context.eventPayload.reason} 
-                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring} 
+                this.logger.debug(`disconnected after unexpected close ${context.eventPayload.reason}
+                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring}
                   and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
                 // this transition circumvents the 'disconnecting' state (since the websocket is already closed),
                 // so we need to execute its onExit behavior here.
@@ -234,7 +241,10 @@ export class RTMClient extends EventEmitter {
       // reconnecting is just like disconnecting, except that the websocket should already be closed before we enter
       // this state, and that the next state should be connecting.
       .state('reconnecting')
-        .do(() => Promise.resolve(true))
+        .do(() => {
+          this.keepAlive.stop();
+          return Promise.resolve(true);
+        })
           .onSuccess().transitionTo('connecting')
         .onExit(() => this.teardownWebsocket())
       .global()

--- a/src/RTMClient.ts
+++ b/src/RTMClient.ts
@@ -155,15 +155,15 @@ export class RTMClient extends EventEmitter {
             .on('websocket close')
               .transitionTo('reconnecting').withCondition(() => this.autoReconnect)
               .withAction((_from, _to, context) => {
-                this.logger.debug(`reconnecting after unexpected close ${context.eventPayload.reason}
-                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring}
-                  and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
+                this.logger.debug(`reconnecting after unexpected close ${context.eventPayload.reason} ` +
+                  `${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring} ` +
+                  `and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
               })
               .transitionTo('disconnected')
               .withAction((_from, _to, context) => {
-                this.logger.debug(`disconnected after unexpected close ${context.eventPayload.reason}
-                  ${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring}
-                  and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
+                this.logger.debug(`disconnected after unexpected close ${context.eventPayload.reason} ` +
+                  `${context.eventPayload.code} with isMonitoring set to ${this.keepAlive.isMonitoring} ` +
+                  `and recommendReconnect set to ${this.keepAlive.recommendReconnect}`);
                 // this transition circumvents the 'disconnecting' state (since the websocket is already closed),
                 // so we need to execute its onExit behavior here.
                 this.teardownWebsocket();

